Add examples to Exception object and group members

janekdb opened the pull request:
- Extend main comment with additional examples
- Group methods from a user perspective
- List exceptions special cased by shouldRethrow
- Include overlooked withTry in opt, either notes.

janekdb commented:
![scaladoc-exception-examples-groups-v1](https://cloud.githubusercontent.com/assets/1123855/14717370/1c533e6a-07e9-11e6-8d57-badb34f1454a.png)

janekdb commented:
Review by @dickwall, @som-snytt, @SethTisue, @parambirs .

som-snytt commented on src/library/scala/util/control/Exception.scala:
```diff
@@ -13,24 +13,125 @@ package control
+ *  This class differs from [[scala.util.Try]] in that it focuses on composing exception handlers rather than
+ *  composing behavior.  All behavior should be composed first and fed to a [[Catch]] object using one of the
+ *  `opt`, `either` or `withTry` methods. Taken together the classes provide a DSL for composing catch and finally
+ *  behaviours.
```
Two flavo(u)rs of behavio(u)r. Where is spell bot? Does it standardise on Swiss spelling? Standardize?

janekdb replied:
Occurrences of each variation,

| Variation | Count |
| --- | --- |
| behavior | 176 |
| behaviour | 51 |

Majority rule => behavior

janekdb replied:
Fixed.

som-snytt replied:
variacioun. http://quod.lib.umich.edu/cgi/m/mec/med-idx?type=id&id=MED50741

som-snytt commented on src/library/scala/util/control/Exception.scala:
```diff
@@ -13,24 +13,125 @@ package control
  *
  *  val s = "http://www.scala-lang.org/"
+ *
+ *  // Option[URL]
```
I think it's more natural to show types inline. `val x: Option[URL] = RHS`. or show a REPL transcript if desirable to show values. I don't know that there is a standard for docs yet, let alone executable examples, but I like the idea of REPL transcripts. I've seen them somewhere. I just saw example at JavaConverters will shows explicit types.

janekdb replied:
I too have read a suggestion that examples should be copy/paste/runnable in the REPL. I started with this approach and I liked it but using an IDEA worksheet was easier in the end. I'll try it again now the examples are stable and see what the output looks like. Failing that moving the types inline is a good idea.

janekdb replied:
This is the raw REPL output from evaluating the first example expression by expression,

```
scala> import scala.util.control.Exception._
import scala.util.control.Exception._

scala> import java.net._
import java.net._

scala> val s = "http://www.scala-lang.org/"
s: String = http://www.scala-lang.org/

scala> val x1 = catching(classOf[MalformedURLException]) opt new URL(s)
x1: Option[java.net.URL] = Some(http://www.scala-lang.org/)

scala> val x2 = catching(classOf[MalformedURLException], classOf[NullPointerException]) either new URL(s)
x2: scala.util.Either[Throwable,java.net.URL] = Right(http://www.scala-lang.org/)

scala> val x3 = catching(classOf[MalformedURLException], classOf[NullPointerException]) withTry new URL(s)
x3: scala.util.Try[java.net.URL] = Success(http://www.scala-lang.org/)

scala> val defaultUrl = new URL("http://example.com")
defaultUrl: java.net.URL = http://example.com

scala> val x4 = failAsValue(classOf[MalformedURLException])(defaultUrl)(new URL("htt/xx"))
x4: java.net.URL = http://example.com
```

The `scala>` prompt is clutter that distracts from the example code but the REPL presentation of the expression values is an improvement over the content of the comment style though I still prefer the comments. I like the idea of augmenting the current comments with the full expression value while including the type explicitly on the LHS.

som-snytt commented on src/library/scala/util/control/Exception.scala:
```diff
@@ -13,24 +13,125 @@ package control
+ *  @groupprio logic-container 60
+ *  @groupdesc logic-container Containers for catch and finally behaviour.
  *
- *  @author Paul Phillips
```
I don't normally like author tags, but I think this one is justified. This stuff is especially paulpy. It's up to the build not emit the doc tags. And other major contributors do add their names at times, I guess for self-branding purposes.

janekdb replied:
Restored.

som-snytt commented:
Thanks, this is educational. At lunch break, I'm going to work through the examples.

janekdb committed:
Add examples to Exception object and group members

 - Extend main comment with additional examples
 - Group methods from a user perspective
 - List exceptions special cased by shouldRethrow
 - Include overlooked withTry in opt, either notes.

janekdb commented:
Review by @som-snytt.

lrytz commented:
LGTM, very nice!

som-snytt commented:
There's also https://issues.scala-lang.org/browse/SI-9073

I don't know why you don't like birds. Why would you want to kill two with one stone?

janekdb commented:
Hey, I like Kestrels in any combination!
